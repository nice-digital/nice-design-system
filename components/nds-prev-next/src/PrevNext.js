import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import "./../scss/prev-next.scss";

const Link = ({ text, destination, elementType: LinkTag = "a", intro }) => {
	let linkProps = {
		className: "prev-next__link"
	};
	if (LinkTag === "a") {
		linkProps.href = destination;
	} else {
		linkProps.to = destination;
	}
	return (
		<LinkTag {...linkProps}>
			<span className="prev-next__link-intro">{intro}</span>
			<span className="prev-next__link-text">{text}</span>
		</LinkTag>
	);
};

export const PrevNext = props => {
	const { nextPageLink, previousPageLink, className, ...rest } = props;

	const nextLinkProps = {
		intro: nextPageLink?.intro || "Next page",
		text: nextPageLink?.text,
		destination: nextPageLink?.destination,
		elementType: nextPageLink?.elementType
	};

	const previousLinkProps = {
		intro: previousPageLink?.intro || "Previous page",
		text: previousPageLink?.text,
		destination: previousPageLink?.destination,
		elementType: previousPageLink?.elementType
	};

	return (
		<div className={classnames("prev-next", className)} {...rest}>
			{(nextPageLink || previousPageLink) && (
				<nav aria-label="Previous and next pages">
					{nextPageLink && <Link {...nextLinkProps} />}
					{previousPageLink && <Link {...previousLinkProps} />}
				</nav>
			)}
		</div>
	);
};

PrevNext.propTypes = {
	nextPageLink: PropTypes.objectOf(Link),
	previousPageLink: PropTypes.objectOf(Link),
	className: PropTypes.string
};

Link.propTypes = {
	text: PropTypes.string.isRequired,
	destination: PropTypes.string.isRequired,
	elementType: PropTypes.elementType,
	intro: PropTypes.string
};
