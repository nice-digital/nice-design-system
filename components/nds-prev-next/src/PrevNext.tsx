import React from "react";
import classnames from "classnames";
import "./../scss/prev-next.scss";

export interface PrevNextLink {
	text: string;
	destination: string;
	elementType?: React.ElementType;
	intro?: string;
	method?: string;
}

export interface PrevNextProps {
	[prop: string]: unknown;
	previousPageLink?: PrevNextLink;
	nextPageLink?: PrevNextLink;
	className?: string;
}

const Link: React.FC<PrevNextLink> = ({
	text,
	destination,
	elementType: ElementType = "a",
	intro,
	method
}: PrevNextLink) => {
	let linkProps = {
		className: "prev-next__link",
		[method || (ElementType === "a" && "href") || "to"]: destination
	};
	return (
		<ElementType {...linkProps}>
			<span className="prev-next__link-intro">{intro}</span>
			<span className="prev-next__link-text">{text}</span>
		</ElementType>
	);
};

export const PrevNext: React.FC<PrevNextProps> = (props: PrevNextProps) => {
	const { nextPageLink, previousPageLink, className, ...rest } = props;

	let nextLinkProps: PrevNextLink | null,
		previousLinkProps: PrevNextLink | null;

	nextLinkProps = nextPageLink
		? {
				intro: nextPageLink?.intro || "Next page",
				text: nextPageLink?.text,
				destination: nextPageLink?.destination,
				elementType: nextPageLink?.elementType,
				method: nextPageLink?.method
		  }
		: null;

	previousLinkProps = previousPageLink
		? {
				intro: previousPageLink?.intro || "Previous page",
				text: previousPageLink?.text,
				destination: previousPageLink?.destination,
				elementType: previousPageLink?.elementType,
				method: previousPageLink?.method
		  }
		: null;

	return (
		<div className={classnames("prev-next", className)} {...rest}>
			{(nextPageLink || previousPageLink) && (
				<nav aria-label="Previous and next pages">
					{nextPageLink && nextLinkProps && <Link {...nextLinkProps} />}
					{previousPageLink && previousLinkProps && (
						<Link {...previousLinkProps} />
					)}
				</nav>
			)}
		</div>
	);
};
