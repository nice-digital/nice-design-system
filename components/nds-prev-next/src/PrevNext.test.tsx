import React from "react";
import { render } from "@testing-library/react";
import { PrevNext } from "./PrevNext";

const Link = () => <a href="#">I am a link</a>;

const nextPageLink = {
	text: "Page Header",
	destination: "/pageheader",
	elementType: Link,
	intro: "Show me the next one"
};

const previousPageLink = {
	text: "Alert",
	destination: "/alert"
};

describe("Previous & Next", () => {
	it("should match snapshot with both links supplied", () => {
		const wrapper = render(
			<PrevNext
				nextPageLink={nextPageLink}
				previousPageLink={previousPageLink}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot with one link supplied", () => {
		const wrapper = render(<PrevNext previousPageLink={previousPageLink} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("link should render an href attribute if the elementType is an anchor", () => {
		const wrapper = render(<PrevNext previousPageLink={previousPageLink} />);
		const anchor = wrapper.getByRole("link");
		expect(anchor.getAttribute("href")).toBe("/alert");
	});

	it("should pass extra props to the containing element", () => {
		const { container } = render(
			<PrevNext nextPageLink={nextPageLink} data-tracker="my-tracker" />
		);
		expect(
			container.querySelector(".prev-next")?.getAttribute("data-tracker")
		).toBe("my-tracker");
	});

	it("should merge additional classnames with the component classname", () => {
		const { container } = render(
			<PrevNext nextPageLink={nextPageLink} className="extra-class" />
		);
		expect(container.querySelector(".prev-next")).toHaveClass(
			"prev-next extra-class"
		);
	});
});
