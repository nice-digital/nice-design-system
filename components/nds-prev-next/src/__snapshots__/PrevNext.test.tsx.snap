// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Previous & Next should match snapshot with both links supplied 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="prev-next"
        data-component="prev-next"
      >
        <nav
          aria-label="Previous and next pages"
        >
          <a
            href="#"
          >
            I am a link
          </a>
          <a
            class="prev-next__link"
            href="/alert"
          >
            <span
              class="prev-next__link-intro"
            >
              Previous page
            </span>
            <span
              class="prev-next__link-text"
            >
              Alert
            </span>
          </a>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="prev-next"
      data-component="prev-next"
    >
      <nav
        aria-label="Previous and next pages"
      >
        <a
          href="#"
        >
          I am a link
        </a>
        <a
          class="prev-next__link"
          href="/alert"
        >
          <span
            class="prev-next__link-intro"
          >
            Previous page
          </span>
          <span
            class="prev-next__link-text"
          >
            Alert
          </span>
        </a>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Previous & Next should match snapshot with one link supplied 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="prev-next"
        data-component="prev-next"
      >
        <nav
          aria-label="Previous and next pages"
        >
          <a
            class="prev-next__link"
            href="/alert"
          >
            <span
              class="prev-next__link-intro"
            >
              Previous page
            </span>
            <span
              class="prev-next__link-text"
            >
              Alert
            </span>
          </a>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="prev-next"
      data-component="prev-next"
    >
      <nav
        aria-label="Previous and next pages"
      >
        <a
          class="prev-next__link"
          href="/alert"
        >
          <span
            class="prev-next__link-intro"
          >
            Previous page
          </span>
          <span
            class="prev-next__link-text"
          >
            Alert
          </span>
        </a>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
